### PR TITLE
Refs #6376 -- Added support for custom gettext domains.

### PR DESCRIPTION
**Pull request by R4G3-BABY**
https://code.djangoproject.com/ticket/6376

This PR changes the ``LOCALE_PATHS`` setting to allow usage of custom gettext domains.

### CODE REVIEW

File: django/utils/translation/trans_real.py [hunk 1]
```diff
@@ -116,8 +116,13 @@ def _new_gnu_trans(self, localedir, use_null_fallback=True):
         Using param `use_null_fallback` to avoid confusion with any other
         references to 'fallback'.
         """
+        if isinstance(localedir, (tuple, list)):
+            localedir, domain = localedir
```
Comment (claudep) [inline django/utils/translation/trans_real.py]: Should we catch the exception in the case the list length != 2?

File: docs/ref/settings.txt [hunk 2]
```diff
@@ -1855,6 +1855,10 @@ Here's a sample settings file::
 ``LOCALE_PATHS``
 ----------------
 
+.. versionchanged:: 2.2
+
+    Allowed usage of custom gettext domains.
+
 Default: ``[]`` (Empty list)
 
 A list of directories where Django looks for translation files.
```
Comment (claudep) [inline docs/ref/settings.txt]: A list of directories `or (directory, domain) tuples` ...

File: tests/i18n/locale/pt/LC_MESSAGES/custom_domain.po [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+"Language-Team: LANGUAGE\n"
+"Language: zxx\n"
+"Plural-Forms: nplurals=3; plural=(n==1 ? 0 : n%10>=2 && n%10<=4 && (n%100<10 "
+"|| n%100>=20) ? 1 : 2);\n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+"X-Loco-Source-Locale: en_PT\n"
+"X-Loco-Parser: loco_parse_po\n"
+"X-Generator: Loco https://localise.biz/"
```
Comment (claudep) [inline tests/i18n/locale/pt/LC_MESSAGES/custom_domain.po]: Please simplify the headers to be minimal (no X- headers, pt Language).

File: tests/i18n/locale/pt/LC_MESSAGES/custom_domain.po [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+msgid "This is a test of the custom ugettext domain loading."
+msgstr "Este é um teste do carregamento de domínio personalizado do ugettext."
```
Comment (claudep) [inline tests/i18n/locale/pt/LC_MESSAGES/custom_domain.po]: Replace ugettext by gettext everywhere

File: tests/i18n/tests.py [hunk 5]
```diff
@@ -1415,6 +1415,26 @@ def test_i18n_patterns_returns_list(self):
...
+    @override_settings(
+        LANGUAGE_CODE='pt',
+        LOCALE_PATHS=[
+            os.path.join(here, 'locale'),
+            (os.path.join(here, 'locale'), 'custom_domain')
+        ]
+    )
+    def test_custom_gettext_domains(self):
+        """
+        This test checks if locale files with custom names have been
```
Comment (claudep) [inline tests/i18n/tests.py]: You can start the comment with `Locale files with...`.

File: django/utils/translation/trans_real.py (hunk 1 above)
Comment (R4G3-BABY) [reply django/utils/translation/trans_real.py]: I wasn't sure what to do about that. What should I do after catching the exception?

Comment (claudep) [reply django/utils/translation/trans_real.py]: Generally, for settings-related errors, we raise `ImproperlyConfigured` exceptions. The alternative would be to create a system check to validate the `LOCALE_PATHS` setting at startup.

Comment (R4G3-BABY) [reply django/utils/translation/trans_real.py]: That should be it.

**Comment by R4G3-BABY** [comment]
@charettes just fixed it, thanks.

**Comment by claudep** [comment]
@timgraham or @carltongibson, would be great to plan a final detailed review for this patch.

File: django/utils/translation/trans_real.py [hunk 6]
```diff
@@ -116,8 +116,19 @@ def _new_gnu_trans(self, localedir, use_null_fallback=True):
         Using param `use_null_fallback` to avoid confusion with any other
         references to 'fallback'.
         """
+        if isinstance(localedir, (tuple, list)):
+            try:
+                localedir, domain = localedir
+            except ValueError:
```
Comment (carltongibson) [inline django/utils/translation/trans_real.py]: Query: Did the system checks added not already catch this error? (i.e. do we need to handle it again here?)

File: docs/ref/settings.txt [hunk 7]
```diff
@@ -1855,16 +1855,21 @@ Here's a sample settings file::
 ``LOCALE_PATHS``
 ----------------
 
+.. versionchanged:: 2.2
```
Comment (carltongibson) [inline docs/ref/settings.txt]: 3.0. Should go at bottom of section, not top.

File: docs/releases/2.2.txt [hunk 8]
```diff
@@ -171,7 +171,8 @@ Generic Views
 Internationalization
 ~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* Changed :setting:`LOCALE_PATHS` setting to allow usage of custom gettext
+  domains (:ticket:`6376`).
```
Comment (carltongibson) [inline docs/releases/2.2.txt]: Move to 3.0.txt

File: docs/topics/i18n/translation.txt [hunk 9]
```diff
@@ -2110,9 +2110,13 @@ message file specific to the project you are composing. The choice is yours.
 All message file repositories are structured the same way. They are:
 
 * All paths listed in :setting:`LOCALE_PATHS` in your settings file are
```
Comment (carltongibson) [inline docs/topics/i18n/translation.txt]: "All paths..." needs some qualification now? 

* tuples are searched `language`, `domain`.
* Paths have an implicit domain of `django`

File: tests/check_framework/test_translation.py [hunk 10]
```diff
@@ -1,34 +1,36 @@
...
+    E001, E002, E003, check_setting_language_code, check_setting_locale_paths,
+)
 from django.test import SimpleTestCase, override_settings
 
 
 class TranslationCheckTests(SimpleTestCase):
 
     def test_valid_language_code(self):
         tags = (
-            'en',              # language
```
Comment (timgraham) [inline tests/check_framework/test_translation.py]: Please revert unrelated whitespace changes.

File: docs/releases/2.2.txt (hunk 8 above)
Comment (timgraham) [reply docs/releases/2.2.txt]: Also omit the ticket reference, that's reserved for bug fixes. "The LOCALE_PATHS setting now supports custom gettext domains." (perhaps with a link to an explanation of what functionality custom gettext domains offers -- maybe that's beyond the scope of Django's documentation but as someone who doesn't use translations, the purpose of custom domains doesn't seem to be described anywhere in these changes.)

File: django/core/checks/translation.py [hunk 11]
```diff
@@ -17,3 +27,17 @@ def check_setting_language_code(app_configs, **kwargs):
     if not language_code_re.match(settings.LANGUAGE_CODE):
         return [E001]
     return []
+
+
+@register(Tags.translation)
+def check_setting_locale_paths(app_configs, **kwargs):
+    errors = []
+
```
Comment (timgraham) [inline django/core/checks/translation.py]: Chop the blank lines please.

File: django/core/checks/translation.py [hunk 12]
```diff
@@ -8,6 +8,16 @@
     id='translation.E001',
 )
 
+E002 = Error(
```
Comment (timgraham) [inline django/core/checks/translation.py]: Document these messages in `docs/ref/checks.txt`.

File: django/utils/translation/trans_real.py [hunk 13]
```diff
@@ -116,8 +116,19 @@ def _new_gnu_trans(self, localedir, use_null_fallback=True):
...
+                raise ImproperlyConfigured(
+                    'Passing a {}-tuple to LOCALE_PATHS is not supported. Pass a '
+                    '(directory, domain) 2-tuple instead.'.format(len(localedir))
+                )
+        else:
+            domain = self.domain
+
```
Comment (timgraham) [inline django/utils/translation/trans_real.py]: chop blank line

File: docs/ref/settings.txt [hunk 14]
```diff
@@ -1855,16 +1855,21 @@ Here's a sample settings file::
...
+
+    Allowed usage of custom gettext domains.
+
 Default: ``[]`` (Empty list)
 
-A list of directories where Django looks for translation files.
+A list of directories or ``(directory, domain)`` tuples where Django looks for
+translation files.
 See :ref:`how-django-discovers-translations`.
```
Comment (timgraham) [inline docs/ref/settings.txt]: Wrap this to the previous line.

File: tests/i18n/tests.py [hunk 15]
```diff
@@ -1415,6 +1416,29 @@ def test_i18n_patterns_returns_list(self):
...
+            (os.path.join(here, 'locale'), 'custom_domain')
+        ]
+    )
+    def test_custom_gettext_domains(self):
+        """
+        Locale files with custom names have been successfully loaded.
+
+        The locale files are located in i18n/locale/pt/LC_MESSAGES.
+        """
+        self.assertEqual(gettext('This is a test of the default gettext domain loading.'),
```
Comment (timgraham) [inline tests/i18n/tests.py]: Use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/i18n/tests.py [hunk 16]
```diff
@@ -1415,6 +1416,29 @@ def test_i18n_patterns_returns_list(self):
...
+                         'Este é um teste do carregamento de domínio do gettext padrão.')
+
+        self.assertEqual(gettext('This is a test of the custom gettext domain loading.'),
+                         'Este é um teste do carregamento de domínio personalizado do gettext.')
+
+        with self.settings(LOCALE_PATHS=[('', '', '')]):
+            with self.assertRaises(ImproperlyConfigured):
```
Comment (timgraham) [inline tests/i18n/tests.py]: assertRaisesMessage

File: tests/i18n/tests.py [hunk 17]
```diff
@@ -1415,6 +1416,29 @@ def test_i18n_patterns_returns_list(self):
         with override_settings(USE_I18N=True):
             self.assertIsInstance(i18n_patterns([]), list)
 
+    @override_settings(
+        LANGUAGE_CODE='pt',
+        LOCALE_PATHS=[
+            os.path.join(here, 'locale'),
+            (os.path.join(here, 'locale'), 'custom_domain')
```
Comment (timgraham) [inline tests/i18n/tests.py]: Include a trailing comma in cases like this so that if more items are added later, this line doesn't have to be modified again.

File: tests/check_framework/test_translation.py [hunk 18]
```diff
@@ -39,3 +41,17 @@ def test_invalid_language_code(self):
                 self.assertEqual(result, [E001])
                 self.assertEqual(result[0].id, 'translation.E001')
                 self.assertEqual(result[0].msg, 'You have provided an invalid value for the LANGUAGE_CODE setting.')
+
+    def test_valid_locale_paths(self):
+        with self.settings(LOCALE_PATHS=['directory', ('directory', 'domain'), ['directory', 'domain']]):
+            self.assertEqual(check_setting_locale_paths(None), [])
+
+    def test_invalid_locale_paths(self):
+        with self.settings(LOCALE_PATHS=[True, 'directory', ['directory', 'domain']]):
```
Comment (timgraham) [inline tests/check_framework/test_translation.py]: Please use separate tests for each case since they aren't dependent on one another.

File: docs/topics/i18n/translation.txt (hunk 9 above)
Comment (R4G3-BABY) [reply docs/topics/i18n/translation.txt]: This is explained on the next paragraph.

**Comment by R4G3-BABY** [comment]
Failing CI build doesn't seem to be related to this changes.

**Comment by R4G3-BABY** [comment]
@carltongibson @timgraham Anything else I need to do so this can be merged?

File: django/core/checks/translation.py [hunk 19]
```diff
@@ -59,3 +69,13 @@ def check_language_settings_consistent(app_configs, **kwargs):
...
         return [E004]
     return []
+
+
+@register(Tags.translation)
+def check_setting_locale_paths(app_configs, **kwargs):
+    for path in settings.LOCALE_PATHS:
+        if not isinstance(path, (str, tuple, list)):
+            return [E005]
+        if isinstance(path, (tuple, list)) and len(path) != 2:
```
Comment (pope1ni) [inline django/core/checks/translation.py]: This should check that the values in the tuple are strings.

File: django/core/checks/translation.py [hunk 20]
```diff
@@ -59,3 +69,13 @@ def check_language_settings_consistent(app_configs, **kwargs):
...
+            return [E006]
+    return []
```
Comment (pope1ni) [inline django/core/checks/translation.py]: Rather than falling immediately it would be better to accumulate messages for all incorrect entries and highlight the incorrect values in the message. Look at how `E003` is done for an example.

This allows all errors to be fixed at once instead of the annoying run-fix-run-fix-… cycle

File: docs/topics/i18n/translation.txt [hunk 21]
```diff
@@ -2105,9 +2105,14 @@ message file specific to the project you are composing. The choice is yours.
...
-  searched for ``<language>/LC_MESSAGES/django.(po|mo)``
-* ``$APPPATH/locale/<language>/LC_MESSAGES/django.(po|mo)``
-* ``$PYTHONPATH/django/conf/locale/<language>/LC_MESSAGES/django.(po|mo)``
+  searched for ``<language>/LC_MESSAGES/<domain>.(po|mo)``
+* ``$APPPATH/locale/<language>/LC_MESSAGES/<domain>.(po|mo)``
+* ``$PYTHONPATH/django/conf/locale/<language>/LC_MESSAGES/<domain>.(po|mo)``
+
+By default, these locations are searched for ``django.(po|mo)`` files.
+You can also specify what translation file names to be searched for
+by adding a ``(path, domain)`` tuple to the :setting:`LOCALE_PATHS` setting.
```
Comment (pope1ni) [inline docs/topics/i18n/translation.txt]: I think you want `directory` instead of `path` here for consistency.

File: docs/topics/i18n/translation.txt [hunk 22]
```diff
@@ -2105,9 +2105,14 @@ message file specific to the project you are composing. The choice is yours.
...
+:setting:`INSTALLED_APPS` always default to the ``django`` domain.
```
Comment (pope1ni) [inline docs/topics/i18n/translation.txt]: I'm not sure this last sentence makes sense - what is it trying to say?

Comment (R4G3-BABY) [reply docs/topics/i18n/translation.txt]: What I mean to say is that directories defined inside INSTALLED_APPS will always default to the django domain.
I've changed the message it should be more clear now, if not let me know.

File: tests/check_framework/test_translation.py [hunk 23]
```diff
@@ -84,3 +85,22 @@ def test_inconsistent_language_settings(self):
             self.assertEqual(check_language_settings_consistent(None), [
                 Error(msg, id='translation.E004'),
             ])
+
+    def test_valid_locale_paths(self):
+        with self.settings(LOCALE_PATHS=['directory']):
+            self.assertEqual(check_setting_locale_paths(None), [])
+        with self.settings(LOCALE_PATHS=[('directory', 'domain'), ['directory', 'domain']]):
+            self.assertEqual(check_setting_locale_paths(None), [])
```
Comment (pope1ni) [inline tests/check_framework/test_translation.py]: Please combine all these valid values into one to avoid repetition.

File: tests/check_framework/test_translation.py [hunk 24]
```diff
@@ -84,3 +85,22 @@ def test_inconsistent_language_settings(self):
...
+
+    def test_invalid_locale_paths(self):
+        msg = 'You have provided an invalid value in the LOCALE_PATHS setting: {!r}.'
+        with self.settings(LOCALE_PATHS=[True]):
+            self.assertEqual(check_setting_locale_paths(None), [
+                Error(msg.format(True), id='translation.E005')
+            ])
+        msg = 'You have provided an invalid tuple in the LOCALE_PATHS setting: {!r}.'
+        for path in [('directory',), ('directory', 'domain', ''), (1, 2, 3)]:
```
Comment (pope1ni) [inline tests/check_framework/test_translation.py]: When using a loop, please use `self.subTest()`.

**Comment by felixxm** [comment]
@R4G3-BABY Thanks for this patch :+1: I rebased from master and pushed minor edits.

@claudep Is there anything left for this ticket? (you mentioned to use `Refs` instead of `Fixed`) maybe using custom domains inside `INSTALLED_APPS` :thinking: .

**Comment by felixxm** [comment]
@claudep  :point_up: :pray:

**Comment by claudep** [comment]
Sorry, it's on my TODO, I may have difficulty to review stuff during the week. I'll do my best!

**Comment by felixxm** [comment]
Thanks! np, take your time.

File: docs/topics/i18n/translation.txt [hunk 25]
```diff
@@ -2105,9 +2105,15 @@ message file specific to the project you are composing. The choice is yours.
...
-* ``$PYTHONPATH/django/conf/locale/<language>/LC_MESSAGES/django.(po|mo)``
+  searched for ``<language>/LC_MESSAGES/<domain>.(po|mo)``
+* ``$APPPATH/locale/<language>/LC_MESSAGES/<domain>.(po|mo)``
+* ``$PYTHONPATH/django/conf/locale/<language>/LC_MESSAGES/<domain>.(po|mo)``
+
+By default, these locations are searched for ``django.(po|mo)`` files.
+You can also customize translation file names to be searched by adding
+a ``(directory, domain)`` tuple to the :setting:`LOCALE_PATHS` setting.
+:setting:`LOCALE_PATHS` directories defined inside :setting:`INSTALLED_APPS`
+setting will always default to the ``django`` domain.
```
Comment (claudep) [inline docs/topics/i18n/translation.txt]: Sorry, but I'm not sure to understand the meaning of that last sentence.

**Comment by claudep** [comment]
I think this patch is missing some changes in `makemessages`/`compilemessages` commands code. We might address this in a further PR, but it would be blocking for the next release.

**Comment by felixxm** [comment]
@claudep Thanks for comments. I added two commits with support for custom domains in `compilemessages` (compile custom locales) and `makemessages` (ignore custom locales :thinking: ). Can you take a look?

**Comment by carltongibson** [comment]
Reviewing but, quick comment, in case Claude gets to this _now_: 

> ...`makemessages` (ignore custom locales 🤔 )

I'm a bit _¿Que?_ here. Maybe this is right, but why? 

(So I'm a user. I think "I'll use a custom gettext domain for my translations". But I can't gather them. Do I have to create a `makemessages` override to use this, or what? — Looks like there's some documentation missing, on how to use this new feature.)

**Comment by felixxm** [comment]
Yeah, that's why there is a thinking emoji there :smile: I wasn't sure if that's a correct behavior. I think that `compilemessages` works fine now, I also fixed `watch_for_translation_changes()` failure so it seems that only `makemessages` remains.

File: docs/topics/i18n/translation.txt (hunk 25 above)
Comment (carltongibson) [reply docs/topics/i18n/translation.txt]: Here maybe something like: 

```
`domain` is the filename (minus .mo/.po extension) used for the actual 
translation files within the LC_MESSAGES folder for each locale. `domain` 
defaults to `'django'`, but you may provide a custom domain for your 
application by adding a ``(directory, domain)`` tuple to the 
:setting:`LOCALE_PATHS` setting.
```

> Sorry, but I'm not sure to understand the meaning of that last sentence.

I think this refers to the fact that (as yet) custom domains **will not** work for `INSTALLED_APPS`.

File: docs/topics/i18n/translation.txt [hunk 26]
```diff
@@ -2105,9 +2105,15 @@ message file specific to the project you are composing. The choice is yours.
 All message file repositories are structured the same way. They are:
 
 * All paths listed in :setting:`LOCALE_PATHS` in your settings file are
-  searched for ``<language>/LC_MESSAGES/django.(po|mo)``
-* ``$APPPATH/locale/<language>/LC_MESSAGES/django.(po|mo)``
-* ``$PYTHONPATH/django/conf/locale/<language>/LC_MESSAGES/django.(po|mo)``
+  searched for ``<language>/LC_MESSAGES/<domain>.(po|mo)``
+* ``$APPPATH/locale/<language>/LC_MESSAGES/<domain>.(po|mo)``
```
Comment (carltongibson) [inline docs/topics/i18n/translation.txt]: This change is misleading. `domain` is not customizable for `INSTALLED_APPS`.

File: docs/topics/i18n/translation.txt [hunk 27]
```diff
@@ -2105,9 +2105,15 @@ message file specific to the project you are composing. The choice is yours.
...
+* ``$PYTHONPATH/django/conf/locale/<language>/LC_MESSAGES/<domain>.(po|mo)``
```
Comment (carltongibson) [inline docs/topics/i18n/translation.txt]: `domain` isn't variable here either right?

**Comment by felixxm** [comment]
Closing due to inactivity.
